### ISSUE TITLE
Instruction to Load LoRA in FINETUNING_CLI.md

### ISSUE THREAD

**Issue by chigkim** [senstella/csm-mlx#13]
Could you provide more detail instruction on how to load Full and LoRA finetuned models to [FINETUNING_CLI.md?](https://github.com/senstella/csm-mlx/blob/master/FINETUNING_CLI.md)
`csm-mlx generate --help` doesn't seem to have an option to use finetuned models?
Thanks!

### Full Finetuned Models

Load final_model.safetensors directly into your CSM model for inference.

### LoRA Models

1. Load the original pre-trained model
2. Apply the adapter weights (adapters.safetensors) using adapter_config.json

**Comment by senstella** [comment]
Hello! And yes. I didn't add a feature to run fine-tuned model directly on CLI yet - soon will! In the meantime, you can use the API.

For full-finetuned model, you can just
```py
from mlx_lm.sample_utils import make_sampler
from csm_mlx import CSM, csm_1b, generate

import audiofile
import numpy as np

csm = CSM(csm_1b())
csm.load_weights("./finetuned_weight.safetensors")

audio = generate(
    csm,
    text="Hello from Sesame.",
    speaker=0,
    context=[],
    max_audio_length_ms=10_000,
    sampler=make_sampler(temp=0.8, top_k=50)
)

audiofile.write("./audio.wav", np.asarray(audio), 24000)
```

Just replace the `./finetuned_weight.safetensors` with your fine-tuned weight path, and it should work!

For LoRA tuned model, you can try:
```py
from mlx_lm.sample_utils import make_sampler
from huggingface_hub import hf_hub_download
from csm_mlx import CSM, csm_1b, generate, load_adapters

import audiofile
import numpy as np

csm = CSM(csm_1b())
weight = hf_hub_download(repo_id="senstella/csm-1b-mlx", filename="ckpt.safetensors")
csm.load_weights(weight)

# Load LoRA here!
load_adapters(csm, "./run-path")

audio = generate(
    csm,
    text="Hello from Sesame.",
    speaker=0,
    context=[],
    max_audio_length_ms=10_000,
    sampler=make_sampler(temp=0.8, top_k=50)
)

audiofile.write("./audio.wav", np.asarray(audio), 24000)
```
Replace the `./run-path` with your training path where it contains `adapters.safetensors` and `adapter_config.json` to ensure it works correctly!

---

For extra, if you want to convert the trained model to original PyTorch model compatible, here's a little hacky script I used.
```py
import torch
from safetensors.torch import load_file

tensors = load_file("ckpt.safetensors")
recovered_tensors = {}
for name, value in tensors.items():
    name = (
        name.replace("up_proj", "w3")
             .replace("down_proj", "w2")
             .replace("gate_proj", "w1")
             .replace("self_attn", "attn")
             .replace("norm.weight", "norm.scale")
             .replace("input_layernorm", "sa_norm")
             .replace("post_attention_layernorm", "mlp_norm")
             .replace("o_proj", "output_proj")
    )
    value = value.to(torch.float32)
    recovered_tensors[name] = value
    print(name, value.shape)

torch.save(recovered_tensors, "pytorch-ckpt.pt")
```

And if you ever want to fuse some LoRA to base weight (Assuming it's fine-tuned from the Sesame's CSM):
```py
from pathlib import Path
from csm_mlx import CSM, csm_1b, load_adapters
from huggingface_hub import hf_hub_download
from mlx.utils import tree_flatten, tree_unflatten
from mlx_lm.utils import save_weights

csm = CSM(csm_1b())
weight = hf_hub_download(repo_id="senstella/csm-1b-mlx", filename="ckpt.safetensors")
csm.load_weights(weight)

load_adapters(csm, './run-path')

fused_linears = [
    (n, m.fuse()) for n, m in csm.named_modules() if hasattr(m, "fuse")
]
csm.update_modules(tree_unflatten(fused_linears))

weights = dict(tree_flatten(csm.parameters()))
save_path = Path("./fused-weight")
save_weights(save_path, weights)
```
Of course, replace the `./run-path` and `fused-weight`.

---

I'm planning to add these features to the documentation and update the CLI with fusing capabilities and `generate` command for custom models as soon as I can make time for it. Thank you so much for opening this issue and bringing it to my attention! I'll leave the issue open, so others can see it in the meantime.

**Comment by chigkim** [comment]
Thanks @senstella so much for the clarification!
On a related note, when I run the fine-tuned model, it often generates silence instead of a sentence. For example, if I generate:
"Hello there! This is Conversation Speech Model from Sesame AI. I'm speaker number 1, fine tuned with audio samples. It's nice to meet you."
It would generate silence for the sentence, "I'm speaker 1, fine tuned with audio samples."
I'd appreciate any suggestion.
Thanks!

**Comment by senstella** [comment]
Glad that it ran!

Regarding the silence issue, I believe its instability stems from fine-tuning directly on the base model. While fine-tuning eliminates some of the weird quirks of the base model, others remain - like how the LLM base model easily collapses into repetitive outputs. I think adjusting the hyperparameters and slightly overfitting on training data might help, if the data is clean enough.

On the other hand, I'm trying to implement offline RL methods in this library and try a training run to minimize the possibilities of weird silence in CSM. It might take a few days to a few weeks due to my limited time, but I'll let you know when offline RLs are implemented!

**Comment by chigkim** [comment]
Thanks @senstella again.
Which hyperparameters would you suggest to change and how?
Dataset is recorded professionally, so it's very clean.

**Comment by senstella** [comment]
I've only just played around with small datasets and tiny batch sizes myself, but if your machine's got the RAM for it, try adjusting up that batch size to around `16` or `32` and the increase the epoch to like `2` or `3` with higher learning rate, maybe even assigning slightly stronger value to `first_codebook_weight_multiplier` like `1.3`. If your dataset is massive, then you might have to use bigger LoRA. I definitely recommend trying out multiple combinations, it depends on your dataset's quarks - so what works best will really depends on what you are working with! You can see the options by using `--help` or see the [docs](https://github.com/senstella/csm-mlx/blob/master/FINETUNING_CLI.md) here.

**Comment by chigkim** [comment]
Thanks! I'll try with different combinations. I'm closing this issue.

**Comment by chigkim** [comment]
Oops, sorry. I think you wanted to keep it open for reference for others. I'm reopening. lol

**Comment by chigkim** [comment]
Just to update, I tried batch size to 16, and epoch to like 35, and loss got down to like 2.6. However, it still produces silence for 1/4 sentences.
I'm trying to full fine tune subsets of VCTK, with about 1.7k samples.
Now there seem different full finetuning like sft, dpo, etc. Could you provide differences and how to choose?
Thanks!

**Comment by senstella** [comment]
Epoch 35 and it's still generating silence? That's definitely weird.

For new fine-tuning methods, SFT is the original cross-entropy loss where you just train the model for next-token prediction. DPO and KTO are offline reinforcement learning methods that allow the model to learn what is 'negative' samples. In theory, these methods should allow the model to learn that 'generating silence' is negative. KTO is much more flexible than DPO, and results from LLM research claim that it generalizes better than DPO in most cases. However, please note that I haven't yet verified if my implementations of these two approaches work correctly - once I verify them, I'll offically put them on the docs.
